src/org/traccar/Config.java

```diff
@@ -15,48 +15,55 @@
  */
 package org.traccar;
 
+import org.apache.commons.lang3.StringUtils;
+
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.Enumeration;
 import java.util.Properties;
 
 public class Config {
 
     private final Properties properties = new Properties();
-    private Properties defaultProperties;
 
-    public void load(String file) throws IOException {
+    void load(String file) throws IOException {
         try (InputStream inputStream = new FileInputStream(file)) {
             properties.loadFromXML(inputStream);
         }
 
         String defaultConfigFile = properties.getProperty("config.default");
         if (defaultConfigFile != null) {
             try (InputStream inputStream = new FileInputStream(defaultConfigFile)) {
-                defaultProperties = new Properties();
+                Properties defaultProperties = new Properties();
                 defaultProperties.loadFromXML(inputStream);
+
+                Enumeration props = defaultProperties.propertyNames();
+                while (props.hasMoreElements()) {
+                    String key = (String) props.nextElement();
+                    if (!properties.containsKey(key)) {
+                        properties.setProperty(key, defaultProperties.getProperty(key));
+                    }
+                }
             }
         }
     }
 
     public boolean hasKey(String key) {
-        return properties.containsKey(key) || defaultProperties != null && defaultProperties.containsKey(key);
+        return properties.containsKey(key);
     }
 
     public String getString(String key) {
-        if (properties.containsKey(key) || defaultProperties == null) {
-            return properties.getProperty(key);
-        } else {
-            return defaultProperties.getProperty(key);
+        String envName = key.toUpperCase().replaceAll("\\.", "_");
```

Comment:
Replacing using regular expressions is a pretty heavy operation. I think we should have some global flag to tell server to consider environment variables. Otherwise just use config.

Also, it would be nice to add underscore between words in camel-case parameters. For example `database.ignoreUnknown` should be converted to `DATABASE_IGNORE_UNKNOWN`.